### PR TITLE
[js/webgpu] Add BatchNormalization Op

### PR DESCRIPTION
**Pull request by qjia7**
### Description
This PR adds `BatchNormalization` with `float` support.

Some Todos:
1. all inputs don't have same data type. For example, x/y is float16, but bias/scale is float32 or double.
2. training mode support.

We see many models are using `BatchNormalization` ops. However, due to the missing in jsep, all of them run on cpu, which result very poor performance. With this PR's support, densenet-9 model becomes 20.29 ms from 250.69 ms.

### CODE REVIEW

**Comment by qjia7** [comment]
@guschmue @satyajandhyala @fs-eire Please take a look, thanks!

File: js/web/lib/wasm/jsep/webgpu/ops/batch-norm.ts [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+            cOffset += `let cOffset = ${scale.indicesToOffset('cIndices')};`;
+          }
+        }
+        return cOffset;
+      };
+      const getInferenceModeShaderSource = (helper: ShaderHelper) => `
+  const epsilon = ${epsilon};
+  ${helper.declareVariables(x, scale, bias, inputMean, inputVar, y)}
+  ${helper.mainStart()}
+  ${helper.guardAgainstOutOfBoundsWorkgroupSizes(outputSize)}
```
Comment (satyajandhyala) [inline js/web/lib/wasm/jsep/webgpu/ops/batch-norm.ts:111]: Can we use uniforms for this and any other possible/applicable variables?

Comment (qjia7) [reply js/web/lib/wasm/jsep/webgpu/ops/batch-norm.ts]: Done.

**Review by satyajandhyala — DISMISSED** [review]

File: js/web/lib/wasm/jsep/webgpu/ops/batch-norm.ts [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+import {ComputeContext, ProgramInfo} from '../types';
+
+import {createTensorShapeVariables, enableShapesUniforms, getMaxComponents, inputVariable, outputVariable, ShaderHelper} from './common';
+
+export interface BatchNormAttributes extends AttributeWithCacheKey {
+  readonly epsilon: number;
+  readonly momentum: number;
+  readonly spatial: boolean;
+  readonly trainingMode: boolean;
+  readonly format: 'nhwc'|'nchw';
```
Comment (fs-eire) [inline js/web/lib/wasm/jsep/webgpu/ops/batch-norm.ts:18]: Maybe using upper cases for the format values?

I found that there are a few places for the tensor layout using lower cases ("nhwc") and some others using upper cases ("NHWC").

WebNN is currently using lower cases, and existing WebGPU attributes are using upper cases.

Comment (fs-eire) [reply js/web/lib/wasm/jsep/webgpu/ops/batch-norm.ts:18]: let me know if you think lower case is better, so that I can make a change to make all JS string literal for tensor layout to lower case.

File: js/web/lib/wasm/jsep/webgpu/ops/batch-norm.ts [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+        switch (i % 10) {
+          case 1:
+            throw new Error(`${message}: ${i}st dimension != ${v}`);
+          case 2:
+            throw new Error(`${message}: ${i}nd dimension != ${v}`);
+          case 3:
+            throw new Error(`${message}: ${i}rd dimension != ${v}`);
+          default:
+            throw new Error(`${message}: ${i}th dimension != ${v}`);
+        }
```
Comment (fs-eire) [inline js/web/lib/wasm/jsep/webgpu/ops/batch-norm.ts:50]: perhaps make this simpler... something like `${message}: dim[${i}] do not match` is good enough

File: onnxruntime/core/providers/js/operators/batch_norm.h [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+class BatchNorm final : public JsKernel {
+ public:
+  explicit BatchNorm(const OpKernelInfo& info) : JsKernel(info) {
+    float epsilon = info.GetAttrOrDefault<float>("epsilon", 1e-5);
+    float momentum = info.GetAttrOrDefault<float>("momentum", 0.9);
+    int64_t spatial = info.GetAttrOrDefault<int64_t>("spatial", 1);
+
+    const auto& node = info.node();
+    int opset = node.SinceVersion();
+    int64_t training_mode = opset <= 9 ? info.GetOutputCount() > 1 : info.GetAttrOrDefault<int64_t>("training_mode", 0);
```
Comment (fs-eire) [inline onnxruntime/core/providers/js/operators/batch_norm.h:21]: if we don't support training mode anyway, perhaps fail it here when training mode is ON and no need to pass to JS

File: js/web/lib/wasm/jsep/webgpu/ops/batch-norm.ts [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    createAttributeWithCacheKey(attributes as Omit<BatchNormAttributes, keyof AttributeWithCacheKey>);
+
+export const batchNorm = (context: ComputeContext, attributes: Record<string, unknown>): void => {
+  const {inputs, outputCount} = context;
+  const updatedAttributes = parseBatchNormAttributes({...attributes, outputCount});
+  if (env.webgpu.validateInputContent) {
+    validateInputs(inputs, updatedAttributes);
+  }
+  if (attributes.trainingMode) {
+    throw new Error('BatchNormalization trainingMode is not supported yet.');
```
Comment (qjia7) [inline js/web/lib/wasm/jsep/webgpu/ops/batch-norm.ts:145]: [1]

File: onnxruntime/core/providers/js/operators/batch_norm.h (hunk 4 above)
Comment (qjia7) [reply onnxruntime/core/providers/js/operators/batch_norm.h:21]: I add throw error [1] in js side. It's convenient to only change js side code if we want to support training mode in future.

File: js/web/lib/wasm/jsep/webgpu/ops/batch-norm.ts (hunk 2 above)
Comment (qjia7) [reply js/web/lib/wasm/jsep/webgpu/ops/batch-norm.ts]: Change it to upper case to align with the current webgpu implementation.
In future, I think it's good to unify webgpu and webnn. It seems that all other flags are lower case. Maybe it's better to use lower case. No strong preference myself.

File: js/web/lib/wasm/jsep/webgpu/ops/batch-norm.ts (hunk 3 above)
Comment (qjia7) [reply js/web/lib/wasm/jsep/webgpu/ops/batch-norm.ts]: Done

File: js/web/lib/wasm/jsep/webgpu/ops/batch-norm.ts [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    }
+    expected.forEach((v, i) => {
+      if (v !== actual[i]) {
+        throw new Error(`${message}: dim[${i}] do not match`);
+      }
+    });
+  };
+
+  if (inputs[0].dims.length > 1) {
+    const shape = Object.seal<Record<typeof attributes.format, number[]>>({
```
Comment (fs-eire) [inline js/web/lib/wasm/jsep/webgpu/ops/batch-norm.ts:40]: is there a reason why using `Object.seal()`?

Comment (qjia7) [reply js/web/lib/wasm/jsep/webgpu/ops/batch-norm.ts]: Offline talked about this. Use `? :` instead of `Object.seal()` in case unknowable performance or error.

**Review by fs-eire — APPROVED** [review]

**Review by fs-eire — APPROVED** [review]
